src/api-service/__app__/onefuzzlib/webhooks.py

```diff
@@ -0,0 +1,165 @@
+            logging.info(
+                "webhook message already handled: %s:%s", self.webhook_id, self.event_id
+            )
+            return
+
+        self.try_count += 1
+
+        logging.debug("sending webhook: %s:%s", self.webhook_id, self.event_id)
+        if self.send():
+            self.state = WebhookMessageState.succeeded
+            self.save()
+            logging.info("sent webhook event: %s:%s", self.webhook_id, self.event_id)
+            return
+
+        if self.try_count < MAX_TRIES:
+            self.state = WebhookMessageState.retrying
+            self.save()
+            self.queue_webhook()
+            logging.warning(
+                "sending webhook event failed, re-queued. %s:%s",
+                self.webhook_id,
+                self.event_id,
+            )
+        else:
+            self.state = WebhookMessageState.failed
+            self.save()
+            logging.warning(
+                "sending webhook event failed %d times. %s:%s",
+                self.try_count,
+                self.webhook_id,
+                self.event_id,
+            )
+
+    def send(self) -> bool:
+        webhook = Webhook.get(self.webhook_id)
+        if not webhook:
+            logging.error(
+                "webhook no longer exists: %s:%s", self.webhook_id, self.event_id
+            )
+            return False
+
+        data = WebhookMessage(
+            webhook_id=self.webhook_id,
+            event_id=self.event_id,
+            event_type=self.event_type,
+            event=self.event,
+        ).json()
+
+        try:
+            response = requests.post(webhook.url, json=data)
+            return response.ok
+        except Exception as err:
+            logging.error(
+                "webhook failed with exception: %s:%s - %s",
+                self.webhook_id,
+                self.event_id,
+                err,
+            )
+            return False
+
+    def queue_webhook(self) -> None:
+        obj = WebhookMessageQueueObj(webhook_id=self.webhook_id, event_id=self.event_id)
+
+        if self.state == WebhookMessageState.queued:
+            visibility_timeout = 0
+        elif self.state == WebhookMessageState.retrying:
+            visibility_timeout = 30
+        else:
+            logging.error(
+                "invalid WebhookMessage queue state, not queuing. %s:%s - %s",
+                self.webhook_id,
+                self.event_id,
+                self.state,
+            )
+            return
+
+        queue_object(
+            "webhooks",
+            obj,
+            visibility_timeout=visibility_timeout,
+            account_id=get_func_storage(),
+        )
+
+
+class Webhook(BASE_WEBHOOK, ORMMixin):
+    @classmethod
+    def key_fields(cls) -> Tuple[str, Optional[str]]:
+        return ("webhook_id", "url")
+
+    @classmethod
+    def send_event(cls, event_type: WebhookEventType, event: WebhookEvent) -> None:
```

Comment:
can't we just infer the event_type form the event type ?

Reply:
We can, and it's probably reasonable for the service side to do that.  That said, I wanted to specifically include the event type in the messages that we send.

Reply:
I agree to keep the event type on the message the suggestion is only for the service. basically changing the signature to 
`def send_event(cls, event: WebhookEvent) -> None` and use the type of the event to infer the event type in the message